# Initial Data Understanding

## Loading Data and Libraries

In [35]:
import pandas as pd
import numpy as np
import sys

import seaborn as sns
import matplotlib.pyplot as plt

sys.path.append('../src/')
pd.set_option('display.max_columns', None)

%matplotlib inline
%load_ext autoreload
%autoreload 2

The autoreload extension is already loaded. To reload it, use:
  %reload_ext autoreload


## Análise preliminar dos dados de Quimioterapia e Radioterapia para Câncer de Mama
Neste documento serão tratadas as primeiras identificações de padrão e possíveis incongruências - de acordo com o julgamento do grupo - em relação aos dados fornecidos para Quimioterapia e Radioterapia de pacientes atendidos para câncer de mama.

In [38]:
import preprocess

kwargs = {'error_bad_lines':False, 'sep':',', 'low_memory':False, 'index_col':0, 'skiprows':0}

df_mama_quim = preprocess.read_to_df('Linfomas Quimioterapia SIA-SUS.csv', kwargs)
df_mama_rad = preprocess.read_to_df('Linfomas Radioterapia SIA-SUS.csv', kwargs)

In [39]:
import constants

df_mama_quim.rename(columns=constants.LAYOUT_APAC, inplace=True)
df_mama_rad.rename(columns=constants.LAYOUT_APAC, inplace=True)

### Amostras de dados
Um primeiro panorama de formatos e disposição de ambos datasets.

In [42]:
df_mama_quim.head(5)

dt_process tp_gestao  cod_tp_gestao  cod_est_cnes       num_apac  \
4       201401        EP         120000       2001586  1213200185967   
25      201401        EP         120000       2001586  1213200175561   
27      201401        EP         120000       2001586  1213200175528   
28      201401        EP         120000       2001586  1213200177266   
32      201401        EP         120000       2001586  1213200186275   
34      201401        EP         120000       2001586  1213200179576   
47      201402        EP         120000       2001586  1214200031759   
52      201402        EP         120000       2001586  1214200031693   
76      201402        EP         120000       2001586  1213200179576   
77      201402        EP         120000       2001586  1214200031000   

    dt_atend_paciente  proc_princ_apac  vl_tot_apac_aprov  cod_uf_mun  tp_est  \
4              201401        304060011            1258.64      120040       5   
25             201401        304060011            1258.64      120040       5   
27             201401        304060011            1258.64      120040       5   
28             201401        304060011            1258.64      120040       5   
32             201401        304070017            1700.00      120040       5   
34             201401        304060038            1258.64      120040       5   
47             201402        304070017            1700.00      120040       5   
52             201402        304060011            1258.64      120040       5   
76             201402        304060038            1258.64      120040       5   
77             201402        304070025            1381.76      120040       5   

    tip_prest est_man_ind  cnpj_est_exe  cnpj_mante          num_csn  \
4          40           I  6.360294e+13         NaN  ‚{}{{}ƒ{{„„€   
25         40           I  6.360294e+13         NaN  ƒ„ƒ{{|{|‚|ƒ€   
27         40           I  6.360294e+13         NaN  ƒ„ƒ{{{{€|}{|}   
28         40           I  6.360294e+13         NaN  ƒ„ƒ{{{{{}~}€   
32         40           I  6.360294e+13         NaN  ƒ„ƒ{{{ƒ{ƒƒ{‚‚   
34         40           I  6.360294e+13         NaN  ||}||{„}~{{{{   
47         40           I  6.360294e+13         NaN  ƒ„ƒ{{{ƒ{ƒƒ{‚‚   
52         40           I  6.360294e+13         NaN  ||{|‚~~{{{|   
76         40           I  6.360294e+13         NaN  ||}||{„}~{{{{   
77         40           I  6.360294e+13         NaN  ƒ„ƒ{{}„}ƒ„|~}   

    cod_idade  num_idade sexo  raca_cor  cod_uf_mun_paciente  nac_paciente  \
4           4         20    M         3               120040            10   
25          4         46    M         3               120040            10   
27          4         20    M         3               120040            10   
28          4         61    M         3               120001            10   
32          4         18    F         3               120040            10   
34          4         25    F         3               120040            10   
47          4         18    F         3               120040            10   
52          4         24    F         3               120040            10   
76          4         25    F         3               120040            10   
77          4         19    M         1               130070            10   

    cep_paciente  uf_res_dif_paciente  mun_res_dif_paciente  dt_ini_val  \
4       69900970                    0                     1    20131105   
25      69900970                    0                     1    20131105   
27      69900970                    0                     1    20131105   
28      69945000                    0                     1    20131105   
32      69900970                    0                     1    20131128   
34      69900970                    0                     1    20131205   
47      69900970                    0                     1    20140204   
52      69901024                    0                     1    2014

In [43]:
df_mama_rad.head(10)

dt_process tp_gestao  cod_tp_gestao  cod_est_cnes       num_apac  \
8       201401        EP         120000       2001586  1214200029119   
14      201402        EP         120000       2001586  1214200045201   
21      201402        EP         120000       2001586  1214200029119   
29      201403        EP         120000       2001586  1214200045201   
76      201405        EP         120000       2001586  1214200084944   
77      201405        EP         120000       2001586  1214200085054   
87      201405        EP         120000       2001586  1214200085043   
89      201405        EP         120000       2001586  1214200084944   
94      201405        EP         120000       2001586  1214200084944   
96      201406        EP         120000       2001586  1214200085043   

    dt_atend_paciente  proc_princ_apac  vl_tot_apac_aprov  cod_uf_mun  tp_est  \
8              201401        304010090                614      120040       5   
14             201402        304010090               2098      120040       5   
21             201402        304010090                630      120040       5   
29             201403        304010090               2550      120040       5   
76             201405        304010090                810      120040       5   
77             201405        304010090               1018      120040       5   
87             201405        304010090                810      120040       5   
89             201403        304010090                695      120040       5   
94             201404        304010090                510      120040       5   
96             201406        304010090                510      120040       5   

    tip_prest est_man_ind    cnpj_est_exe  cnpj_mante          num_csn  \
8          40           I  63602940000170         NaN  ƒ„ƒ{{~„€ƒ„ƒ€ƒ   
14         40           I  63602940000170         NaN  ƒ„ƒ{{}„€~~ƒ{}   
21         40           I  63602940000170         NaN  ƒ„ƒ{{~„€ƒ„ƒ€ƒ   
29         40           I  63602940000170         NaN  ƒ„ƒ{{}„€~~ƒ{}   
76         40           I  63602940000170         NaN  }{{€{{{|}{{{~   
77         40           I  63602940000170         NaN  ‚{{{€~ƒ€ƒ{   
87         40           I  63602940000170         NaN  ‚{ƒ€{ƒ~„~|‚€{‚   
89         40           I  63602940000170         NaN  }{{€{{{|}{{{~   
94         40           I  63602940000170         NaN  }{{€{{{|}{{{~   
96         40           I  63602940000170         NaN  ‚{ƒ€{ƒ~„~|‚€{‚   

    cod_idade  num_idade sexo  raca_cor  cod_uf_mun_paciente  nac_paciente  \
8           4          8    M         3               120050            10   
14          4         75    M         3               120040            10   
21          4          8    M         3               120050            10   
29          4         75    M         3               120040            10   
76          4         13    M         3               120020            10   
77          4         33    M         3               120040            10   
87          4         72    M         3               120040            10   
89          4         13    M         3               120020            10   
94          4         13    M         3               120020            10   
96          4         72    M         3               120040            10   

    cep_paciente  uf_res_dif_paciente  mun_res_dif_paciente  dt_ini_val  \
8       69940000                    0                     1    20140123   
14      69900970                    0                     1    20140212   
21      69940000                    0                     1    20140123   
29      69900970                    0                     1    20140212   
76      69980000                    0                     1    20140318   
77      69900970                    0                     1    20140514   
87      69900970                    0                     1    20140514   
89      69980000                    0        

### Geração de reports
Mesmo com análises exaustivas de frações dos dados, um overview total - de correlações, missing values et al - é imprescindível. Utilizando `pandas_profiling`, geraremos então relatórios que servirão de insumo do cenário macro desses dados.

In [40]:
from pandas_profiling import ProfileReport

ProfileReport(df_mama_quim).to_file('mama_quimioterapia_report.html')
ProfileReport(df_mama_rad).to_file('mama_radioterapia_report.html')

### Missing values

Devemos nos atentar ao valor informativo das colunas - porcentagens muito altas de missing values são indicativos de inaplicabilidade ou simplesmente não preenchimento da informação.

Primeiramente, nos dados de quimioterapia:

In [64]:
1 - df_mama_quim.count().sort_values().head(20) / len(df_mama_quim)

etnia                0.999716
aq_cidini3           0.973344
aq_cidini2           0.937684
aq_dt_ini_3_trat     0.927430
dt_ocorr             0.901655
aq_dt_ini_2_trat     0.887407
aq_cidini1           0.835039
aq_dt_ini_1_trat     0.771918
aq_cid_10            0.670408
aq_esu_p2            0.472046
cnpj_mante           0.271705
num_apac_ant         0.095159
dt_auto              0.075321
aq_estadi            0.041121
dt_soli              0.039926
cod_soli_cnes        0.003130
aq_grahis            0.000527
cnpj_est_exe         0.000035
aq_linfo_reg_inva    0.000000
aq_dt_inden_pato     0.000000
dtype: float64

A porcentagem de valores faltantes em `etnia` é certamente alarmante, uma coluna praticamente 100% nula. Vejamos seus poucos não-nulos

In [68]:
len(df_mama_quim[~df_mama_quim.etnia.isnull()]), len(df_mama_quim)

(90, 316965)

Seguidamente, nos dados de radioterapia:

In [69]:
1 - df_mama_rad.count().sort_values().head(20) / len(df_mama_rad)

ar_smrd             1.000000
etnia               0.999046
ar_cidini3          0.976004
ar_cid_topo_3       0.947691
ar_cidini2          0.943122
ar_dt_ini_3_trat    0.940562
ar_dt_fim_3         0.937952
ar_dt_ini_3         0.937952
ar_dt_ini_2_trat    0.896436
ar_cid_topo_2       0.839408
ar_dt_fim_2         0.830572
ar_dt_ini_2         0.830572
ar_cidini1          0.717871
ar_dt_ini_1_trat    0.659036
ar_num_ins_3        0.562801
dt_ocorr            0.561697
ar_num_ins_2        0.506928
ar_cid_10           0.464056
cnpj_mante          0.359689
ar_trat_anter       0.284137
dtype: float64

O cenário de 99.9%+ nulos observados em `etnia` se repetem também aqui, porém a `ar_smrd` - que inclusive não possui descrição no dicionário correlato - se apresenta com absolutos 100%. Iremos portanto descartar esta coluna.

In [70]:
df_mama_rad = df_mama_rad.drop(columns=['ar_smrd'])

### Valores constantes

De modo similar a missing values, colunas com valores constantes também não apresentam valor significativo. A partir do relatório, notamos essa característica no **Tipo de Atendimento APAC** para ambos sets.

In [114]:
df_mama_quim.tp_atend_apac.unique()

array([3], dtype=int64)

In [115]:
df_mama_rad.tp_atend_apac.unique()

array([4], dtype=int64)

Portanto, serão desconsideradas como colunas.

In [121]:
df_mama_quim = df_mama_quim.drop(columns=['tp_atend_apac'])
df_mama_rad = df_mama_rad.drop(columns=['tp_atend_apac'])

### Agrupamento de datas
Diversas colunas em ambos sets são marcações de tempo, que podem ajudar a inferir a duração dum tratamento.

Como primeiro passo para tratá-las, devemos convertê-las ao formato `datetime`.

In [127]:
dtcols = [c for c in df_mama_quim.columns if c.startswith('dt_')]
df_mama_quim[dtcols].head(10)

dt_process  dt_atend_paciente  dt_ini_val  dt_fim_val  dt_ocorr  \
4       201401             201401    20131105    20140131       NaN   
25      201401             201401    20131105    20140131       NaN   
27      201401             201401    20131105    20140131       NaN   
28      201401             201401    20131105    20140131       NaN   
32      201401             201401    20131128    20140131       NaN   
34      201401             201401    20131205    20140228       NaN   
47      201402             201402    20140204    20140430       NaN   
52      201402             201402    20140207    20140430       NaN   
76      201402             201402    20131205    20140228       NaN   
77      201402             201402    20140218    20140430       NaN   

       dt_soli   dt_auto  
4   20131105.0  20131105  
25  20131105.0  20131105  
27  20131105.0  20131105  
28  20131105.0  20131105  
32  20131128.0  20131128  
34  20131205.0  20131205  
47  20140204.0  20140204  
52  20140207.0  20140207  
76  20131205.0  20131205  
77  20140218.0  20140218

In [176]:
df_mama_quim[dtcols].dtypes

dt_process           period[M]
dt_atend_paciente    period[M]
dt_ini_val               int64
dt_fim_val               int64
dt_ocorr               float64
dt_soli                float64
dt_auto                 object
dtype: object

As duas primeiras colunas - mês de processamento e atendimento ao paciente - formatam-se em ano-mês, portanto terão conversão diferenciada

In [150]:
ymth_dt = [dtcols[0:2]]
to_year_month = lambda df: df.apply(pd.to_datetime, format='%Y%M').apply(lambda date: date.to_period('M'))

In [144]:
df_mama_quim[ymth_dt] = df_mama_quim[ymth_dt].apply(to_year_month)
df_mama_rad[ymth_dt] = df_mama_rad[ymth_dt].apply(to_year_month)

df_mama_quim[ymth_dt].head(5)

dt_process dt_atend_paciente
4     2014-01           2014-01
25    2014-01           2014-01
27    2014-01           2014-01
28    2014-01           2014-01
32    2014-01           2014-01

Nota-se que `dt_soli` (data de solicitação) e `dt_ocorr` (data de ocorrência) são colunas `float`, portanto contando com casas decimais que não fazem sentido para o formato `%Y%M%d`.

In [206]:
ymd_dt = [dtcols[2:-1]]
# Retirada de casas decimais e conversão para datetime
to_ymd = lambda df: df.fillna(0).astype(int).astype(str).replace('0', np.nan).apply(pd.to_datetime, format='%Y%M%d')

In [207]:
df_mama_rad[['dt_ocorr']].apply(to_ymd)

dt_ocorr
8                      NaT
14                     NaT
21     2014-01-28 00:02:00
29     2014-01-31 00:03:00
76     2014-01-31 00:05:00
77                     NaT
87                     NaT
89                     NaT
94                     NaT
96     2014-01-16 00:06:00
102    2014-01-30 00:06:00
124                    NaT
130                    NaT
132    2014-01-31 00:07:00
156    2014-01-30 00:07:00
231                    NaT
267                    NaT
271    2015-01-31 00:12:00
353                    NaT
362                    NaT
366    2016-01-16 00:05:00
373    2016-01-31 00:05:00
387                    NaT
403    2016-01-01 00:06:00
426    2016-01-31 00:07:00
467                    NaT
674                    NaT
686                    NaT
696                    NaT
802    2014-01-15 00:02:00
...                    ...
744946 2018-01-30 00:11:00
744959                 NaT
744964                 NaT
744979                 NaT
744980                 NaT
745003 2018-01-22 00:11:00
745014 2018-01-20 00:12:00
745026 2018-01-21 00:12:00
745065                 NaT
745401                 NaT
745445                 NaT
745465                 NaT
745478 2014-01-01 00:06:00
745520 2014-01-01 00:07:00
745524                 NaT
745541                 NaT
745554 2014-01-31 00:07:00
745565                 NaT
745584                 NaT
745588 2014-01-31 00:08:00
745603 2014-01-31 00:08:00
745610 2014-01-31 00:08:00
745635                 NaT
745649 2014-01-30 00:09:00
745677                 NaT
745683 2014-01-30 00:09:00
745736 2014-01-31 00:10:00
745738 2014-01-31 00:10:00
745745 2014-01-31 00:10:00
746029 2018-01-07 00:11:00

[19920 rows x 1 columns]